File: web/src/application.zig
const std = @import("std");

const js = struct {
    extern fn webCanvasInit() void;
    extern fn webCanvasDeinit() void;
    extern fn webCanvasSetSize(width: u32, height: u32) void;
    extern fn webCanvasSetTitle(title: [*]const u8, len: u32) void;
};

pub const Hints = struct {};

pub fn init(width: u32, height: u32, title: []const u8, hints: Hints) void {
    js.webCanvasInit();

    setSize(width, height);
    setTitle(title);

    _ = hints;
}

pub fn deinit() void {
    js.webCanvasDeinit();
}

pub fn setSize(width: u32, height: u32) void {
    js.webCanvasSetSize(width, height);
}

pub fn setTitle(title: []const u8) void {
    js.webCanvasSetTitle(title.ptr, title.len);
}

File: web/src/application.js
const original_title = document.title;
const text_decoder = new TextDecoder();
let log_buf = "";

const web = {
	canvas: 0,
	wasm: undefined,

	init: function(wasm_engine) {
		web.wasm = wasm_engine;
	},

	getString: function(str, len) {
		const memory = web.wasm.exports.memory.buffer;
		return text_decoder.decode(new Uint8Array(memory, str, len));
	},

	webLogWrite: function(str, len) {
		log_buf += web.getString(str, len);
	},

	webLogFlush: function() {
		console.log(log_buf);
		log_buf = "";
	},

	webPanic: function(str, len) {
		throw Error(web.getString(str, len));
	},

	webCanvasInit: function() {
		web.canvas = document.createElement("canvas");
		document.body.appendChild(web.canvas);
	},

	webCanvasDeinit: function() { },

	webCanvasSetSize: function(width, height) {
		web.canvas.width = width;
		web.canvas.heigh = height;
	},

	webCanvasSetTitle: function(str, len) {
		const title = len > 0 ? web.getString(str, len) : original_title;
		document.title = title;
	},
};

export { web }
